pandas/_libs/tslibs/conversion.pyx

```diff
@@ -0,0 +1,499 @@
+            one_diff = np.where(np.diff(trans_idx) != 1)[0] +1
+            trans_grp = np.array_split(trans_idx, one_diff)
+
+            # Iterate through each day, if there are no hours where the
+            # delta is negative (indicates a repeat of hour) the switch
+            # cannot be inferred
+            for grp in trans_grp:
+
+                delta = np.diff(result_a[grp])
+                if grp.size == 1 or np.all(delta > 0):
+                    stamp = _render_tstamp(vals[grp[0]])
+                    raise pytz.AmbiguousTimeError(stamp)
+
+                # Find the index for the switch and pull from a for dst and b
+                # for standard
+                switch_idx = (delta <= 0).nonzero()[0]
+                if switch_idx.size > 1:
+                    raise pytz.AmbiguousTimeError(
+                        "There are %i dst switches when "
+                        "there should only be 1." % switch_idx.size)
+                switch_idx = switch_idx[0] + 1 # Pull the only index and adjust
+                a_idx = grp[:switch_idx]
+                b_idx = grp[switch_idx:]
+                dst_hours[grp] = np.hstack((result_a[a_idx], result_b[b_idx]))
+
+    for i in range(n):
+        left = result_a[i]
+        right = result_b[i]
+        if vals[i] == NPY_NAT:
+            result[i] = vals[i]
+        elif left != NPY_NAT and right != NPY_NAT:
+            if left == right:
+                result[i] = left
+            else:
+                if infer_dst and dst_hours[i] != NPY_NAT:
+                    result[i] = dst_hours[i]
+                elif is_dst:
+                    if ambiguous_array[i]:
+                        result[i] = left
+                    else:
+                        result[i] = right
+                elif fill:
+                    result[i] = NPY_NAT
+                else:
+                    stamp = _render_tstamp(vals[i])
+                    raise pytz.AmbiguousTimeError(
+                        "Cannot infer dst time from %r, try using the "
+                        "'ambiguous' argument" % stamp)
+        elif left != NPY_NAT:
+            result[i] = left
+        elif right != NPY_NAT:
+            result[i] = right
+        else:
+            if is_coerce:
+                result[i] = NPY_NAT
+            else:
+                stamp = _render_tstamp(vals[i])
+                raise pytz.NonExistentTimeError(stamp)
+
+    return result
+
+
+cdef inline bisect_right_i8(int64_t *data, int64_t val, Py_ssize_t n):
+    cdef Py_ssize_t pivot, left = 0, right = n
+
+    # edge cases
+    if val > data[n - 1]:
+        return n
+
+    if val < data[0]:
+        return 0
+
+    while left < right:
+        pivot = left + (right - left) // 2
+
+        if data[pivot] <= val:
+            left = pivot + 1
+        else:
+            right = pivot
+
+    return left
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline str _render_tstamp(int64_t val):
```

Comment:
so instead of doing this, why don't you simply do

```
from pandas._libs.tslib import Timestamp
return str(Timestamp(val))
```

of course this is a local import and only occurs when you are raising anyhow.

much more obvious

Reply:
I thought about this and couldn't fight off the allure of self-contained-ness.  Will change.